src/drivers/intel/cavs/timer.c

```diff
@@ -36,22 +36,55 @@ void platform_timer_stop(struct timer *timer)
 		   shim_read(SHIM_DSPWCTCS) & ~SHIM_DSPWCTCS_T0A);
 }
 
+/* clock ticks it takes to set clock and enable IRQ */
+#define TIMER_OVERHEAD	500
+
+/* max time it should take to do 2 subsequenmt timer reads */
+#define TIMER_READ_TIME	256
+
+static spinlock_t timer_lock = {0};
+
 int64_t platform_timer_set(struct timer *timer, uint64_t ticks)
 {
+	uint64_t ticks_now[2];
+	uint32_t flags;
+
 	/* a tick value of 0 will not generate an IRQ */
 	if (ticks == 0)
 		ticks = 1;
 
-	/* Check if requested time is not past time */
-	if (ticks > shim_read64(SHIM_DSPWC))
+	spin_lock_irq(&timer_lock, flags);
+
+	/* 64bit reads are non atomic on xtensa so we must
+	 * read a stable value where there is no bit 32 flipping.
+	 * A large delta between reads[0..1] means we have flipped
+	 * and that the value read back in either 0..1] is invalid.
+	 */
+	do {
+		ticks_now[0] = shim_read64(SHIM_DSPWC);
+		ticks_now[1] = shim_read64(SHIM_DSPWC);
+
+		/* worst case is we perform this twice so 4 clock reads */
+	} while (ticks_now[1] - ticks_now[0] > TIMER_READ_TIME);
```

Comment:
@andyross fyi

Reply:
Got it.  FWIW, you can do a little better by reading the 32 bit subwords independently and testing not on the 64 bit delta but on the identity of high word, in pseudocode:

```c
uint32_t hi0, hi1, lo;
uint64_t ticks_now;
do {
    hi0 = read32(DSPWC_HI);
    lo = read32(DSPWC_LO);
    hi1 = read32(DSPWC_HI);
} while (hi0 != hi1);
ticks_now = (((uint64_t)hi0) << 32) | lo;
```

That comes to a worst case of 6 MMIO register reads instead of 8, and replaces the 64 subtract and compare with a single 32 bit equality test.

Reply:
Good point @andyross , will update.